doc/rfc/781-lua-system-wide-profiler.md

```diff
@@ -0,0 +1,395 @@
+** guests stacks in format described in <lj_sysprof.h>
+*/
+#define LUAM_SYSPROF_LEAF 1
+/*
+** CALLGRAPH mode = DEFAULT + streams samples with full callchains of host
+** and guest stacks in format described in <lj_sysprof.h>
+*/
+#define LUAM_SYSPROF_CALLGRAPH 2
+```
+
+
+Available profiler options are stored in the `luam_sysprof_options` structure.
+
+```c
+/* Profiler options. */
+struct luam_Sysprof_Options {
+  /* Profiling mode. */
+  uint8_t mode;
+  /* Sampling interval in msec. */
+  uint64_t interval;
+  /* Custom buffer to write data. */
+  uint8_t *buf;
+  /* The buffer's size. */
+  size_t len;
+  /* Context for the profile writer and final callback. */
+  void *ctx;
+};
+```
+
+As you can see, the writer backend is customizable, and our implementation does not lock it at the C-API level.
+It is required to provide a writer, a backtracer, and an on-stop callback before running the profiler. If one wants to use the default backtracer, he can call `luaM_sysprof_set_backtracer` with `NULL` argument.
+To start the profiler, one needs to fill in the `luam_Sysprof_Options` with desired values and pass it to the `luaM_sysprof_start` function. In the case of a successful start, it returns `PROFILE_SUCCESS`.
+
+
+To stop the profiler, one should call `luaM_sysprof_stop` function, which will return `PROFILE_SUCCESS` in case of successful profiling finalization.
+
+Starting the profiler in Lua is simple:
+
+```lua
+1  local started, err, errno = misc.sysprof.start{
+2    mode = profiling_mode,
+3    interval = sampling_interval,
+4    path = fname,
+5  }
+```
+
+where `profiling_mode` is one of the following:
+```
+"D" - DEFAULT mode
+"L" - LEAF mode
+"C" - CALLGRAPH mode
+```
+
+In the `DEFAULT` mode, only the counters for each state will be collected. In the `LEAF` mode, the sysprof provides you with information about a top stack frame for each sample. And in the `CALLCHAIN` mode, the profiler gives you a complete call stack for each sample.
+
+Take a look at the examples below to understand the difference.
+| ![](https://i.imgur.com/4J9ItbN.png) |
+|:---:|
+|`DEFAULT` mode |
+
+
+| ![Leaf mode](https://i.imgur.com/ALIVNn1.png) |
+|:--:|
+|`LEAF` mode|
+
+| ![](https://i.imgur.com/X0nL83r.png) |
+| :--: |
+| `CALLCHAIN` mode |
+
+
+The `sampling_interval` is the sampling period in milliseconds. `path` is name of the file where profile events are written. This function's writer calls `fwrite()` for each call, retrying if EINTR occurs. When the profiling is stopped, `fclose()` is called. The contract for the profiler function is similar to that of standard io* interfaces. If it is impossible to open a file for writing or the profiler fails to start, nil is returned on failure (plus an error message as a second result and a system-dependent error code as a third result). Otherwise, it returns the `true` value.
+
+Stopping profiler from Lua is simple too:
+```lua
+1  local stopped, err, errno = misc.sysprof.stop()
```

Comment:
As far as we don't reference this code we can omit line numbers.
Feel free to ignore.